plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc

```diff
@@ -0,0 +1,225 @@
+        (uniform ExtendedOBJMaterial*)material;
+
+    vec3f Kd = make_vec3f(0.f);
+    if (!objMaterial)
+        Kd = make_vec3f(dg.color);
+    else
+        foreach_unique(mat in objMaterial) Kd = mat->Kd * make_vec3f(dg.color);
+
+    Ray localray = ray;
+    vec3f accucolor = make_vec3f(0.f);
+    vec3f mask = make_vec3f(1.f);
+
+    varying RandomTEA* uniform rng =
+        createRandomNumberGenerator(&self->abstract.super, sample);
+
+    // path tracing loop
+    for (int bounces = 0; bounces < NB_MAX_PATH_TRACING_REBOUNDS; bounces++)
+    {
+        traceRay(self->abstract.super.model, localray);
+
+        // if ray misses scene (no hit occurs), return background colour
+        if (localray.geomID < 0)
+        {
+            const vec3f bgcol =
+                make_vec3f(skyboxMapping((Renderer*)self, localray,
+                                         self->abstract.bgMaterial)) *
+                DEFAULT_SKY_POWER;
+
+            accucolor = accucolor + mask * bgcol;
+            break;
+        }
+        DifferentialGeometry dg;
+        postIntersect(self->abstract.super.model, dg, localray,
+                      DG_NG | DG_NS | DG_NORMALIZE | DG_FACEFORWARD |
+                          DG_MATERIALID | DG_COLOR);
+
+        uniform Material* material = dg.material;
+        uniform ExtendedOBJMaterial* objMaterial =
+            (uniform ExtendedOBJMaterial*)material;
+        if (!objMaterial)
+            Kd = make_vec3f(dg.color);
+        else
+            foreach_unique(mat in objMaterial) Kd =
+                mat->Kd * make_vec3f(dg.color);
+
+        // compute ray direction of uniformly random diffuse ray
+        const vec3f raydir =
+            getRandomVector(sample, dg.Ns, self->abstract.super.epsilon);
+
+        // origin of new ray in path is hitpoint of previous ray in path
+        const vec3f hitpoint = dg.P + self->abstract.super.epsilon * dg.Ns;
+        setRay(localray, hitpoint, raydir);
+        localray.t0 = self->abstract.super.epsilon;
+        localray.t = self->abstract.ambientOcclusionDistance;
+
+        if (self->abstract.shadows > 0.f)
+        {
+            // Direct lighting
+            for (uniform int i = 0;
+                 self->abstract.lights && i < self->abstract.numLights; ++i)
+            {
+                const uniform Light* uniform light = self->abstract.lights[i];
+                const vec2f s = make_vec2f(0.5f);
+                const varying Light_SampleRes lightSample =
+                    light->sample(light, dg, s);
+                const vec3f radiance = lightSample.weight;
+                if (reduce_max(radiance) > 0.f)
+                {
+                    const vec3f lightDirection = lightSample.dir;
+                    const vec3f sunSampleDirection =
+                        getConeSample(lightDirection, rng,
+                                      self->abstract.softShadows);
+                    const float sunLight =
```

Comment:
I would rename sunLight to sunAngle to avoid confusion and reflect better what it's doing (that was a bad naming choice on my part, sorry)

Reply:
Sure, I usually use cosNL for diffuse and cosLR for specular. In this case, only diffuse shading is processed. And Sun is probably a specific case too since we can have many lights. I'll change it to cosNL then, unless you have an objection.